interpreter/cling/lib/Interpreter/Value.cpp

```diff
@@ -135,15 +135,53 @@ namespace {
 namespace cling {
 
   Value::Value(const Value& other):
-    m_Storage(other.m_Storage), m_StorageType(other.m_StorageType),
+    m_Storage(other.m_Storage), m_NeedsManagedAlloc(other.m_NeedsManagedAlloc),
+    m_TypeKind(other.m_TypeKind),
     m_Type(other.m_Type), m_Interpreter(other.m_Interpreter) {
     if (other.needsManagedAllocation())
       AllocatedValue::getFromPayload(m_Storage.m_Ptr)->Retain();
   }
 
+  static bool needsManagedAlloc(clang::QualType QT) {
+    clang::QualType Canon = QT.getCanonicalType();
+    if (Canon->isPointerType() || Canon->isObjectType() ||
+        Canon->isReferenceType())
+      if (Canon->isRecordType() || Canon->isConstantArrayType() ||
+          Canon->isMemberPointerType())
+        return true;
+    return false;
+  }
+
+  static Value::TypeKind getCorrespondingTypeKind(clang::QualType QT,
+                                                  const clang::ASTContext &C) {
+    using namespace clang;
+    if (const auto *ET = dyn_cast<EnumType>(QT.getTypePtr()))
+      QT = ET->getDecl()->getIntegerType();
+
+    if (C.hasSameType(QT, C.VoidTy))
+      return Value::kVoid;
+
+    if (!QT->isBuiltinType() || QT->castAs<BuiltinType>()->isNullPtrType())
+      return Value::kPtrTy;
+
+    switch(QT->getAs<BuiltinType>()->getKind()) {
+    default:
+#ifndef NDEBUG
+      QT->dump();
+#endif // NDEBUG
+      assert(false && "Type not supported");
+      return Value::kInvalid;
+#define X(type, name) \
+      case BuiltinType::name: return Value::k##name;
+      BUILTIN_TYPES
+#undef X
+    }
+  }
+
   Value::Value(clang::QualType clangTy, Interpreter& Interp):
-    m_StorageType(determineStorageType(clangTy)),
-    m_Type(clangTy.getAsOpaquePtr()),
+    m_NeedsManagedAlloc(needsManagedAlloc(clangTy)),
+    m_TypeKind(getCorrespondingTypeKind(clangTy, Interp.getCI()->getASTContext())),
```

Comment:
You might be able to save some cycle for inverting the order of calls so that `getCorrespondingTypeKind` is first and can be used as input to `needsManagedAlloc`:
```
  static bool needsManagedAlloc(clang::QualType QT, TypeKind typeKind) {
    if (typeKind == Value::kPtrTy) {
      clang::QualType Canon = QT.getCanonicalType();
      ...
    }
    return false;
  }
```